File: solutions/python/high-scores/1/high_scores.py
from dataclasses import dataclass


@dataclass
class HighScores:
    scores: list[int]

    def latest(self) -> int:
        return self.scores[-1]

    def personal_best(self) -> int:
        return max(self.scores)

    def personal_top_three(self) -> list[int]:
        return sorted(self.scores, reverse=True)[:3]

File: solutions/cpp/high-scores/1/high_scores.cpp
#include "high_scores.h"

#include <numeric>

namespace arcade {

using namespace std;

vector<int> HighScores::list_scores() { return scores; }

int HighScores::latest_score() { return scores.back(); }

int HighScores::personal_best() {
    return accumulate(scores.begin(), scores.end(), 0,
                      [](auto a, auto b) { return max(a, b); });
}

vector<int> HighScores::top_three() {
    vector<int> top;
    auto scores_copy = scores;
    make_heap(scores_copy.begin(), scores_copy.end());
    for (int i = 0; i < 3 && !scores_copy.empty(); ++i) {
        pop_heap(scores_copy.begin(), scores_copy.end());
        top.push_back(scores_copy.back());
        scores_copy.pop_back();
    }
    return top;
}

}  // namespace arcade
